File: c++/Fuerza bruta.cpp
// tsp_brute_force.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <chrono>

using namespace std;
using namespace std::chrono;

vector<pair<double, double>> cities = {
    {0, 0}, {1, 5}, {5, 2}, {3, 8}, {7, 6},
    {8, 3}, {6, 1}, {2, 2}, {4, 7}, {9, 5}
};

double distance(pair<double, double> a, pair<double, double> b) {
    return hypot(a.first - b.first, a.second - b.second);
}

double total_distance(const vector<int>& route) {
    double dist = 0;
    for (int i = 0; i < route.size() - 1; i++)
        dist += distance(cities[route[i]], cities[route[i+1]]);
    dist += distance(cities[route.back()], cities[route[0]]);
    return dist;
}

int main() {
    vector<int> route;
    for (int i = 0; i < cities.size(); i++) route.push_back(i);

    double best_dist = 1e9;
    vector<int> best_route;

    auto start = high_resolution_clock::now();

    do {
        double d = total_distance(route);
        if (d < best_dist) {
            best_dist = d;
            best_route = route;
        }
    } while (next_permutation(route.begin(), route.end()));

    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    cout << "Tiempo de ejecución: " << elapsed.count() << " segundos" << endl;
    return 0;
}

File: Python/Fuerza bruta.py
# tsp_brute_force.py
import itertools
import time
import math

# Coordenadas fijas de 10 ciudades
cities = [
    (0, 0), (1, 5), (5, 2), (3, 8), (7, 6),
    (8, 3), (6, 1), (2, 2), (4, 7), (9, 5)
]

def distance(a, b):
    return math.hypot(a[0] - b[0], a[1] - b[1])

def total_distance(route):
    return sum(distance(cities[route[i]], cities[route[i+1]]) for i in range(len(route)-1)) + distance(cities[route[-1]], cities[route[0]])

def tsp_brute_force():
    best_dist = float('inf')
    best_route = None
    for perm in itertools.permutations(range(len(cities))):
        dist = total_distance(perm)
        if dist < best_dist:
            best_dist = dist
            best_route = perm
    return best_dist, best_route

start_time = time.time()
dist, route = tsp_brute_force()
end_time = time.time()

print(f"Tiempo de ejecución: {end_time - start_time:.6f} segundos")
